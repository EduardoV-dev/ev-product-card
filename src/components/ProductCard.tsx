import React, { createContext, CSSProperties } from 'react';
import {
  IInitialValues,
  IOnChangeArgs,
  IProductCardHandlers,
  IProductContextProps,
  Product,
} from '../interfaces/interfaces';
import useProduct from '../hooks/useProduct';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as IProductContextProps);
const { Provider } = ProductContext;

export interface IProps {
  // children?: ReactElement | ReactElement[];
  children: (args: IProductCardHandlers) => JSX.Element;
  className?: string;
  initialValues?: IInitialValues;
  onChange?: (args: IOnChangeArgs) => void;
  product: Product;
  style?: CSSProperties;
  value?: number;
}

const ProductCard = ({
  children,
  className,
  initialValues,
  onChange,
  product,
  style,
  value,
}: IProps): JSX.Element => {
  const { counter, increaseBy, isMaxCountReached, reset } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount: initialValues?.maxCount,
      }}
    >
      <div
        className={`${styles.productCard} ${className ?? ''}`}
        {...{ style }}
      >
        {children({
          counter,
          isMaxCountReached,

          maxCount: initialValues?.maxCount || Infinity,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};

export default ProductCard;
