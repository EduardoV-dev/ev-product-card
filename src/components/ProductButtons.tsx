import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProps {
  className?: string;
  style?: CSSProperties;
}

const ProductButtons = ({ className = '', style }: IProps) => {
  const { counter, increaseBy, maxCount = Infinity } = useContext(
    ProductContext
  );

  const isMaxReached: () => boolean = useCallback(() => counter >= maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} {...{ style }}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        {' '}
        -{' '}
      </button>
      <div className={styles.countLabel}> {counter} </div>

      <button
        className={`${styles.buttonAdd} ${
          isMaxReached() ? styles.disabled : ''
        }`}
        disabled={isMaxReached()}
        onClick={() => increaseBy(1)}
      >
        {' '}
        +{' '}
      </button>
    </div>
  );
};

export default ProductButtons;
