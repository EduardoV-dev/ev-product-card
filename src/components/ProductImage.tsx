import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProps {
  className?: string;
  img?: string;
  style?: CSSProperties;
}

const ProductImage = ({ className = '', img, style }: IProps) => {
  const { product } = useContext(ProductContext);

  const imgToShow: string = img ?? (product.img = '');

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Coffee Mug"
      {...{ style }}
    />
  );
};

export default ProductImage;
