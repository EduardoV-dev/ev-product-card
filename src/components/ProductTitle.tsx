import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProps {
  className?: string;
  style?: CSSProperties;
  title?: string;
}

const ProductTitle = ({ title, className = '', style }: IProps) => {
  const { product } = useContext(ProductContext);

  return (
    <span
      className={`${styles.productDescription} ${className}`}
      {...{ style }}
    >
      {title ?? product.title}
    </span>
  );
};

export default ProductTitle;
