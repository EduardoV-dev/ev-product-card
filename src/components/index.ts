import { IProductCardHOCProps } from '../interfaces/interfaces';
import ProductButtons from './ProductButtons';
import ProductCardHOC from './ProductCard';
import ProductImage from './ProductImage';
import ProductTitle from './ProductTitle';

export { default as ProductTitle } from './ProductTitle';
export { default as ProductImage } from './ProductImage';
export { default as ProductButtons } from './ProductButtons';

export const ProductCard: IProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export default ProductCard;
