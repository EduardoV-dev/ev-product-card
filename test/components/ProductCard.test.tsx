import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../fixtures';

describe('<ProductCard />', () => {
  it('Should render correctly', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => <ProductTitle title="Custom Product" />}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('Should increment the count', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {({ counter, increaseBy }) => (
          <>
            <h1>Product Card</h1>
            <span>{counter}</span>
            <button onClick={() => increaseBy(1)}>Increase</button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();
    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
