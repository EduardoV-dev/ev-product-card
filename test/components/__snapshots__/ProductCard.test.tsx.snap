// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> Should increment the count 1`] = `
<div
  className="productCard "
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    Increase
  </button>
</div>
`;

exports[`<ProductCard /> Should render correctly 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Custom Product
  </span>
</div>
`;
